### PR TITLE
Psd 393/add new attribution error

### PR DESCRIPTION
**Pull request by iamnotbrandonchan**
This PR addresses PSD-393:
- We want a new error case for the instance that multiple attribution sources fail
- We don't want to use AssetNotFoundError, as it's name does not imply the new error case we are seeing within our code base
- We have a new error case that our handle logs, one for multiple attribution failures

### CODE REVIEW

File: pkg/domain/attributor.go [hunk 1]
```diff
@@ -141,6 +141,20 @@ func (err AssetInventoryMultipleAssetsFoundError) Error() string {
...
 
+// AssetInventoryMultipleAttributionErrors occurs when multiple attribution errors
+// occur on multiple attribution sources
+type AssetInventoryMultipleAttributionErrors struct {
+	Inner         error
+	AssetID       string
+	ScanTimestamp string
+}
+
+func (err AssetInventoryMultipleAttributionErrors) Error() string {
```
Comment (faldridge) [inline pkg/domain/attributor.go]: Suggestion: Similar to the change for PSD-191, I'd like for us to move toward a static `Reason` field, and we can log the other pieces, e.g., `AssetID`, `ScanTimestamp`, and even `Inner` as additional fields in the structured error log line emitted from the handler code.

Comment (iamnotbrandonchan) [reply pkg/domain/attributor.go]: done! err... resolved

File: pkg/logs/assetattributor_error.go [hunk 2]
```diff
@@ -32,3 +34,29 @@ type UnknownAttributionFailureError struct {
...
+	Reason  string `logevent:"reason,default=asset-could-not-attribute-on-sources"`
+	AssetID int64  `logevent:"assetID,default=id-not-specified"`
+}
+
+// AttributionErrorLogFactory is a factory function that takes an error that occurs during attribution,
+// and returns a corresponding struct with logging information
+func AttributionErrorLogFactory(attributionErr error, assetID int64) interface{} {
+	switch attributionErr.(type) {
+	case domain.AssetNotFoundError:
+		return AssetNotFoundError{Reason: attributionErr.Error(), AssetID: assetID}
```
Comment (faldridge) [inline pkg/logs/assetattributor_error.go]: Task: This most recent change did decouple the error business logic from the log handling in the handler, as we discussed, but it did not convert the `Reason` field to static messages as requested. Please update each error's `Error` method to return its `Reason` field value without the dynamic values.

Comment (iamnotbrandonchan) [reply pkg/logs/assetattributor_error.go]: Wait but why would we not want the dynamic values? I kinda want to refute this, because I feel like we should give a detailed error message, even if it doesn't have dynamic values embedded

Comment (iamnotbrandonchan) [reply pkg/logs/assetattributor_error.go]: when you say "reason" field, you mean `Error`?

File: gateway-incoming.yaml [hunk 3]
```diff
@@ -95,6 +95,7 @@ paths:
               "status":
               #! if eq .Response.Body.errorType "AssetNotFoundError" !# 404,
               #! else if eq .Response.Body.errorType "AssetInventoryRequestError" !# 502,
+              #! else if eq .Response.Body.errorType "AssetInventoryMultipleAttributionErrors" !# 404,
```
Comment (mikerott) [inline gateway-incoming.yaml]: Is this really supposed to be 404?  Seems like something in the 5xx range would be more appropriate.

File: pkg/assetattributor/cloudassetinventory.go [hunk 4]
```diff
@@ -92,8 +92,6 @@ func (n *CloudAssetInventory) Attribute(ctx context.Context, asset domain.Nexpos
 	if asset.IP == "" && asset.Hostname == "" {
 		return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetNotFoundError{
 			Inner:          fmt.Errorf("asset has no IP or hostname"),
-			AssetID:        fmt.Sprintf("%d", asset.ID),
```
Comment (mikerott) [inline pkg/assetattributor/cloudassetinventory.go]: Please use `strconv.Itoa` function instead of `fmt.Sprintf`. https://golang.org/pkg/strconv/#Itoa

File: pkg/assetattributor/cloudassetinventory.go [hunk 5]
```diff
@@ -141,29 +139,22 @@ func (n *CloudAssetInventory) Attribute(ctx context.Context, asset domain.Nexpos
 		case httpMultipleAssetsFoundError:
 			return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetInventoryMultipleAssetsFoundError{
 				Inner:          e,
-				AssetID:        fmt.Sprintf("%d", asset.ID),
```
Comment (mikerott) [inline pkg/assetattributor/cloudassetinventory.go]: Use strconv.Itoa

File: pkg/assetattributor/cloudassetinventory.go [hunk 6]
```diff
@@ -141,29 +139,22 @@ func (n *CloudAssetInventory) Attribute(ctx context.Context, asset domain.Nexpos
...
-				ScanTimestamp:  asset.ScanTime.Format(time.RFC3339Nano),
 				AssetInventory: cloudAssetInventoryIdentifier,
 			}
 		case httpBadRequest:
 			return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetInventoryRequestError{
 				Inner:          e,
-				AssetID:        fmt.Sprintf("%d", asset.ID),
```
Comment (mikerott) [inline pkg/assetattributor/cloudassetinventory.go]: Use strconv.Itoa

File: pkg/domain/attributor.go [hunk 7]
```diff
@@ -99,46 +99,52 @@ func (n *CloudAssetDetails) UnmarshalJSON(data []byte) error {
...
-	AssetID        string
-	ScanTimestamp  string
 	AssetInventory string
 }
 
 func (err AssetNotFoundError) Error() string {
 	return fmt.Sprintf(
-		"Result not found for asset with ID %s as of scan time %s in asset inventory %s: %v",
-		err.AssetID, err.ScanTimestamp, err.AssetInventory, err.Inner)
+		"Result not found for asset asset inventory %s: %v",
```
Comment (mikerott) [inline pkg/domain/attributor.go]: There's a grammatical error here I think.  What is the error message supposed to say?

File: gateway-incoming.yaml (hunk 3 above)
Comment (iamnotbrandonchan) [reply gateway-incoming.yaml]: maybe like a 502?

File: pkg/assetattributor/cloudassetinventory.go (hunk 4 above)
Comment (iamnotbrandonchan) [reply pkg/assetattributor/cloudassetinventory.go]: this line is deleted :/

File: pkg/assetattributor/cloudassetinventory.go (hunk 5 above)
Comment (iamnotbrandonchan) [reply pkg/assetattributor/cloudassetinventory.go]: deleted doh

File: pkg/assetattributor/cloudassetinventory.go (hunk 6 above)
Comment (iamnotbrandonchan) [reply pkg/assetattributor/cloudassetinventory.go]: same as the previous

File: pkg/domain/attributor.go (hunk 7 above)
Comment (iamnotbrandonchan) [reply pkg/domain/attributor.go]: yea... this makes 0 sense. fixed!

File: pkg/logs/assetattributor_error.go (hunk 2 above)
Comment (faldridge) [reply pkg/logs/assetattributor_error.go]: Well, what I mean is: we use structured logging, where every field passed to the `logger.Error` interface function is recorded as a dedicated field in the JSON-encoded string. So, we get the static (searchable, groupable) message for Reason that is more specific than the static message given by the `Message` field, but also get the dynamic values in a way that can be further grouped and analyzed by our logging platform when troubleshooting issues.

So, as long as we pass these dynamic fields to the function in the handler, we get the best of both worlds.

Comment (iamnotbrandonchan) [reply pkg/logs/assetattributor_error.go]: yea i deleted the timestamp field to make this more static

File: pkg/domain/attributor.go [hunk 8]
```diff
@@ -99,46 +98,50 @@ func (n *CloudAssetDetails) UnmarshalJSON(data []byte) error {
...
 	AssetInventory string
+	Code           int
+	ScanTimestamp  string
 }
 
 func (err AssetNotFoundError) Error() string {
-	return fmt.Sprintf(
-		"Result not found for asset with ID %s as of scan time %s in asset inventory %s: %v",
-		err.AssetID, err.ScanTimestamp, err.AssetInventory, err.Inner)
+	return "Result not found for asset using asset inventory"
```
Comment (faldridge) [inline pkg/domain/attributor.go]: Suggestion: how does the string "result not found in asset inventory" sound instead? Also, for best sorting and grouping results, we should probably follow some rules like "all lower case, no punctuation" for these types of strings.

File: pkg/domain/attributor.go [hunk 9]
```diff
@@ -99,46 +98,50 @@ func (n *CloudAssetDetails) UnmarshalJSON(data []byte) error {
...
 	AssetInventory string
 	Code           int
+	ScanTimestamp  string
 }
 
 func (err AssetInventoryRequestError) Error() string {
-	return fmt.Sprintf(
-		"Request to asset inventory %s failed with code %d for asset with ID %s as of scan time %s: %v",
-		err.AssetInventory, err.Code, err.AssetID, err.ScanTimestamp, err.Inner)
+	return "Request to asset inventory for asset"
```
Comment (faldridge) [inline pkg/domain/attributor.go]: Suggestion: swap out this string for "request to asset inventory failed"

File: pkg/domain/attributor.go [hunk 10]
```diff
@@ -99,46 +98,50 @@ func (n *CloudAssetDetails) UnmarshalJSON(data []byte) error {
...
-	ScanTimestamp  string
 	AssetInventory string
+	ScanTimestamp  string
 }
 
 func (err AssetInventoryMultipleAssetsFoundError) Error() string {
-	return fmt.Sprintf(
-		"Request to asset inventory %s returned multiple values for asset with ID %s as of scan time %s: %v",
-		err.AssetInventory, err.AssetID, err.ScanTimestamp, err.Inner)
+	return "Request to asset inventory %s returned multiple values for asset"
```
Comment (faldridge) [inline pkg/domain/attributor.go]: Task: remove the unused formatting directive here (`%s`).
Suggestion: make "Request" lowercase; otherwise, I like the text here.

File: pkg/logs/assetattributor_error.go [hunk 11]
```diff
@@ -32,3 +34,33 @@ type UnknownAttributionFailureError struct {
...
+	case *domain.AssetInventoryMultipleAssetsFoundError:
+		attributionErr := attributionErr.(domain.AssetInventoryMultipleAssetsFoundError)
+		return AssetInventoryMultipleAssetsFoundError{Message: attributionErr.Error(), Reason: attributionErr.Inner.Error(), AssetID: assetID}
+	case *domain.AssetInventoryMultipleAttributionErrors:
+		attributionErr := attributionErr.(domain.AssetInventoryMultipleAttributionErrors)
+		return AssetInventoryMultipleAttributionErrors{Message: attributionErr.Error(), Reason: attributionErr.Inner.Error(), AssetID: assetID}
+	default:
+		return UnknownAttributionFailureError{Message: attributionErr.Error(), AssetID: assetID}
+	}
+}
```
Comment (faldridge) [inline pkg/logs/assetattributor_error.go]: ❤️

File: pkg/assetattributor/cloudassetinventory.go [hunk 12]
```diff
@@ -85,16 +85,12 @@ type CloudAssetInventory struct {
 // Attribute queries the asecurityteam/asset-inventory-api service first by IP, then by hostname
 // if the first query returns no results
 func (n *CloudAssetInventory) Attribute(ctx context.Context, asset domain.NexposeAssetVulnerabilities) (domain.NexposeAttributedAssetVulnerabilities, error) {
-	if asset.ScanTime.IsZero() {
```
Comment (faldridge) [inline pkg/assetattributor/cloudassetinventory.go]: Question: why remove this check?

Comment (iamnotbrandonchan) [reply pkg/assetattributor/cloudassetinventory.go]: because asset-producer guarantees this field to be nonzero

Comment (mikerott) [reply pkg/assetattributor/cloudassetinventory.go]: You should keep the check though.  This service (asset-attributor) should not assume all communication originates from asset-producer, nor should its defense of the payload contract be reduced.

Comment (iamnotbrandonchan) [reply pkg/assetattributor/cloudassetinventory.go]: i'll put it back

Comment (faldridge) [reply pkg/assetattributor/cloudassetinventory.go]: There a variety of factors in play here. I fully agree with @mikerott that no code should ever be written with assumptions about what happens on the other side of a service boundary, both to ensure program behavioral correctness, and for maintainability of the code.

However! The OpenAPI spec for this attribute is required to be a non-empty string in RFC 3339 datetime format, and so this much at least is actually enforced at this service's boundary, and so it would seem that the check here is redundant.

And yet! The thrilling drama concludes with this: in go, `time.IsZero` does not check for a value of empty string, but for the zero value of a `time` instance, which is `0001-01-01 00:00:00+0000 UTC`. This means that the zero value - which is absolutely an invalid value that should be rejected - could be validated by the transportd gateway and reach this code path, making the check necessary.

TL;DR: the check **should** remain, but it's important that everyone understand why.
